### DISCUSSION TITLE
Getting errors about including imgui.h when it should just work fine

### DISCUSSION THREAD

**Discussion by nathangur** [ocornut/imgui#5981] [New Users Build/Link/Run issues ONLY!]
Hey all. First time diving into ImGUI and I'm getting a bunch of errors in my main access script (everything is in c++ by the way.) 

Here is that script for reference: 

`
  // PhysicsTest.cpp
  
  #include "simulation.h"
  #include "imgui/imgui.h"
  #include "SDL/include/SDL_events.h"
  #include "SDL/include/SDL_opengl.h"
  #include "SDL/include/SDL.h"
  #include "SDL/include/SDL_syswm.h"
  
  // Simulation and window variables
  PhysicsSimulation simulation;
  SDL_Window* window = nullptr;
  SDL_GLContext gl_context = nullptr;
  
  // Update function that is called every frame
  void update() {
      // Calculate the time elapsed between each frame
      float deltaTime = ImGui::GetIO().DeltaTime;
  
      // Check if the simulation is paused
      if (!simulation.isPaused()) {
          // If the simulation is not paused, simulate different types of motion based on user input
          if (ImGui::RadioButton("Constant Velocity", simulation.getCurrentMotion() == 0)) {
              simulation.setCurrentMotion(0);
          }
          if (ImGui::RadioButton("Gravity", simulation.getCurrentMotion() == 1)) {
              simulation.setCurrentMotion(1);
          }
          if (ImGui::RadioButton("Elastic Collision", simulation.getCurrentMotion() == 2)) {
              simulation.setCurrentMotion(2);
          }
  
          switch (simulation.getCurrentMotion()) {
          case 0:
              simulation.constantVelocity(deltaTime); // Simulate constant velocity
              break;
          case 1:
              simulation.gravity(deltaTime); // Simulate gravity
              break;
          case 2:
              simulation.elasticCollision(deltaTime); // Simulate elastic collision
              break;
          }
  
          simulation.collisionDetection(); // Detect collisions between objects
          simulation.collisionResolution(); // Resolve all collisions
      }
  
      simulation.draw(); // Draw objects on the application window
      simulation.handleInput(); // Handle user input
  }
  
  int main() {
      // Initialize SDL
      if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_TIMER | SDL_INIT_GAMECONTROLLER) != 0) {
          printf("Error: %s\n", SDL_GetError());
          return -1;
      }
  
      // Set up SDL window
      SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);
      SDL_GL_SetAttribute(SDL_GL_DEPTH_SIZE, 24);
      SDL_GL_SetAttribute(SDL_GL_STENCIL_SIZE, 8);
      SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
      // Set up SDL window (cont.)
      SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 2);
      SDL_WindowFlags window_flags = (SDL_WindowFlags)(SDL_WINDOW_OPENGL | SDL_WINDOW_RESIZABLE | SDL_WINDOW_ALLOW_HIGHDPI);
      window = SDL_CreateWindow("Physics Simulation", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, 1280, 720, window_flags);
      gl_context = SDL_GL_CreateContext(window);
      // Initialize objects
      std::vector<Object> objects = {
          {1, 0, 1, 0},
          {2, 1, 2, 0},
          {3, 2, 3, 0},
          {4, 3, 4, 0},
          {5, 4, 5, 0}
      };
      // Create a simulation object with the initialized objects
      simulation = PhysicsSimulation(objects);
  
      // Initialize imgui
      ImGui::CreateContext();
      ImGuiIO& io = ImGui::GetIO(); (void)io;
      ImGui::StyleColorsDark();
  
      // Set up imgui rendering
      ImGui_ImplOpenGL3_Init("#version 150");
      ImGui_ImplSDL2_InitForOpenGL(window, gl_context);
  
      // Main loop
      bool done = false;
      while (!done) {
          // Poll events from SDL
          SDL_Event event;
          while (SDL_PollEvent(&event)) {
              // Pass event data to imgui
              ImGui_ImplSDL2_ProcessEvent(&event);
              // Handle events
              if (event.type == SDL_QUIT) {
                  done = true;
              }
          }
  
          // Start a new imgui frame
          ImGui_ImplOpenGL3_NewFrame();
          ImGui_ImplSDL2_NewFrame(window);
          ImGui::NewFrame();
  
          // Update the simulation and render the GUI
          update();
          ImGui::Render();
          SDL_GL_MakeCurrent(window, gl_context);
          SDL_GL_SwapWindow(window);
      }
  
      // Clean up
      ImGui_ImplOpenGL3_Shutdown();
      ImGui_ImplSDL2_Shutdown();
      ImGui::DestroyContext();
      SDL_GL_DeleteContext(gl_context);
      SDL_DestroyWindow(window);
      SDL_Quit();
  
      return 0;
  }
`

From this script I get the following errors: 

- Severity	Code	Description	Project	File	Line	Suppression State
Error	C2661	'ImGui::RadioButton': no overloaded function takes 1 arguments	23	
Error (active)	E0140	too many arguments in function call		35	
Error (active)	E0140	too many arguments in function call	38	
Error (active)	E0140	too many arguments in function call	41	
Error (active)	E0020	identifier "ImGui_ImplSDL2_Shutdown" is undefined	118	
Error (active)	E0020	identifier "ImGui_ImplSDL2_ProcessEvent" is undefined	97	
Error (active)	E0020	identifier "ImGui_ImplSDL2_NewFrame" is undefined	106	
Error (active)	E0020	identifier "ImGui_ImplSDL2_InitForOpenGL" is undefined	88	
Error (active)	E0020	identifier "ImGui_ImplOpenGL3_Shutdown" is undefined	117	
Error (active)	E0020	identifier "ImGui_ImplOpenGL3_NewFrame" is undefined	105	
Error (active)	E0020	identifier "ImGui_ImplOpenGL3_Init" is undefined	Physics Test	C87	
Error (active)	E0135	class "PhysicsSimulation" has no member "setCurrentMotion"	24	
Error (active)	E0135	class "PhysicsSimulation" has no member "setCurrentMotion"	27	
Error (active)	E0135	class "PhysicsSimulation" has no member "setCurrentMotion"	30	
Error (active)	E0135	class "PhysicsSimulation" has no member "getCurrentMotion"	23	
Error (active)	E0135	class "PhysicsSimulation" has no member "getCurrentMotion"	P26	
Error (active)	E0135	class "PhysicsSimulation" has no member "getCurrentMotion"	29	
Error (active)	E0135	class "PhysicsSimulation" has no member "getCurrentMotion"	33	
Error	C2039	'setCurrentMotion': is not a member of 'PhysicsSimulation'	Physics Test	24	
Error	C2039	'setCurrentMotion': is not a member of 'PhysicsSimulation'	Physics Test	27	
Error	C2039	'setCurrentMotion': is not a member of 'PhysicsSimulation'	Physics Test	30	
Error	C2660	'PhysicsSimulation::gravity': function does not take 1 arguments	38	
Error	C2660	'PhysicsSimulation::elasticCollision': function does not take 1 arguments	41	
Error	C2660	'PhysicsSimulation::constantVelocity': function does not take 1 arguments	35	
Error	C2512	'PhysicsSimulation': no appropriate default constructor available	11	
Error	C3861	'ImGui_ImplSDL2_Shutdown': identifier not found	118	
Error	C3861	'ImGui_ImplSDL2_ProcessEvent': identifier not found	97	
Error	C3861	'ImGui_ImplSDL2_NewFrame': identifier not found	106	
Error	C3861	'ImGui_ImplSDL2_InitForOpenGL': identifier not found	88	
Error	C3861	'ImGui_ImplOpenGL3_Shutdown': identifier not found	117	
Error	C3861	'ImGui_ImplOpenGL3_NewFrame': identifier not found	105	
Error	C3861	'ImGui_ImplOpenGL3_Init': identifier not found	87	
Error	C2661	'ImGui::RadioButton': no overloaded function takes 1 arguments	26	
Error	C2661	'ImGui::RadioButton': no overloaded function takes 1 arguments	29	
Error	C2039	'getCurrentMotion': is not a member of 'PhysicsSimulation'	23	
Error	C2039	'getCurrentMotion': is not a member of 'PhysicsSimulation'	26	
Error	C2039	'getCurrentMotion': is not a member of 'PhysicsSimulation'	29	
Error	C2039	'getCurrentMotion': is not a member of 'PhysicsSimulation'	33	

From all of this there is a bunch of not found identifiers. If I add includes for each of them another issue springs up as it now complains about imgui.h no being found but only on 3 files. Changing it to this: #include "imgui/imgui.h" it fixes itself but then I get an extra 60 errors of random things going haywire. Could anyone assist me in debugging this? 

-Thanks

**Comment by PathogenDavid** [comment]
Hi, please make sure your posts are well-formatted. You need to use three backticks for multi-line codeblocks, IE:

~~~
```cpp
int main()
{
    printf("Hello, world!");
}
```
~~~

becomes

```cpp
int main()
{
    printf("Hello, world!");
}
```

Also the error list shown in the `Output` tab will have much better formatting for copy+pasting.

---------------

You need to include the headers for the backends too. IE: `imgui_impl_sdl.h` and `imgui_impl_opengl3.h`.

You should not need to use the `imgui/` or `imgui/backends/` (or `SDL/include/`) prefixes on your includes. The fact that you do means your project is not set up correctly. Those directories should be added to the include paths of your project at the build system level.

> I get an extra 60 errors of random things going haywire. Could anyone assist me in debugging this?

We can't assist you without knowing what the errors are.

**Comment by LuZixuan2022** [comment]
When I use it, it says:
![image](https://github.com/ocornut/imgui/assets/114159378/adbf5459-7323-45d3-a9ca-d96dc3c988aa)

It went wrong `C2059` in line 2999 at `imgui_internal.h`.

Line `2991`~`3013` in `imgui_internal.h` is like these:
```cpp
namespace ImGui
{
    // Windows
    // We should always have a CurrentWindow in the stack (there is an implicit "Debug" window)
    // If this ever crash because g.CurrentWindow is NULL it means that either
    // - ImGui::NewFrame() has never been called, which is illegal.
    // - You are calling ImGui functions after ImGui::EndFrame()/ImGui::Render() and before the next ImGui::NewFrame(), which is also illegal.
    inline    ImGuiWindow*  GetCurrentWindowRead()      { ImGuiContext& g = *GImGui; return g.CurrentWindow; }
    inline    ImGuiWindow*  GetCurrentWindow()          { ImGuiContext& g = *GImGui; g.CurrentWindow->WriteAccessed = true; return g.CurrentWindow; }
    IMGUI_API ImGuiWindow*  FindWindowByID(ImGuiID id);
    IMGUI_API ImGuiWindow*  FindWindowByName(const char* name);
    IMGUI_API void          UpdateWindowParentAndRootLinks(ImGuiWindow* window, ImGuiWindowFlags flags, ImGuiWindow* parent_window);
    IMGUI_API void          UpdateWindowSkipRefresh(ImGuiWindow* window);
    IMGUI_API ImVec2        CalcWindowNextAutoFitSize(ImGuiWindow* window);
    IMGUI_API bool          IsWindowChildOf(ImGuiWindow* window, ImGuiWindow* potential_parent, bool popup_hierarchy);
    IMGUI_API bool          IsWindowWithinBeginStackOf(ImGuiWindow* window, ImGuiWindow* potential_parent);
    IMGUI_API bool          IsWindowAbove(ImGuiWindow* potential_above, ImGuiWindow* potential_below);
    IMGUI_API bool          IsWindowNavFocusable(ImGuiWindow* window);
    IMGUI_API void          SetWindowPos(ImGuiWindow* window, const ImVec2& pos, ImGuiCond cond = 0);
    IMGUI_API void          SetWindowSize(ImGuiWindow* window, const ImVec2& size, ImGuiCond cond = 0);
    IMGUI_API void          SetWindowCollapsed(ImGuiWindow* window, bool collapsed, ImGuiCond cond = 0);
    IMGUI_API void          SetWindowHitTestHole(ImGuiWindow* window, const ImVec2& pos, const ImVec2& size);
    IMGUI_API void          SetWindowHiddenAndSkipItemsForCurrentFrame(ImGuiWindow* window);
```
